### ISSUE TITLE
ffmpeg > 6.0 should have oneVPL enabled

### ISSUE THREAD

**Issue by septem9er** [NixOS/nixpkgs#303074]
ffmpeg has enabled oneVPL support starting from Version 6.0 ([Changelog](https://git.ffmpeg.org/gitweb/ffmpeg.git/blob/HEAD:/Changelog)).

Therefore I think that the NixOS Package should follow this, and build ffmpeg_6 with:
```
withVpl = true;
withMfx = false; # cannot be enabled together with oneVPL
```

**Comment by septem9er** [comment]
@Atemu
You are the maintainer for the ffmpeg package. What do you think about this?

**Comment by Atemu** [comment]
The problem is not enabling VPL but rather disabling MFX as that would remove support for older intel GPUs.

If there was VPL support for sufficiently old intel GPUs or if ffmpeg could be built with both VPL and MFX at the same time, we could enable VPL. Otherwise, you can always `ffmpeg.override { withVPL = true; withMFX = false; }` for yourself.

cc @evanrichter

**Comment by septem9er** [comment]
I get that disabling MFX is an issue.

Is this change compared to upstream documented somewhere? I couldn't find anything. Could we improve the documentation, so that people know about this change and how to enable libvpl support in ffmpeg without having to look through the nixpkg?

Personally I was configuring hardware acceleration for jellyfin, the wiki of which specifically says that libva is enabled in ffmpeg > 6.0.

That this isn't the case in Nixpkg, not even with ffmpeg-jellyfin, isn't mentioned in the [Jellyfin](https://wiki.nixos.org/wiki/Jellyfin), [Accelerated Video Playback](https://wiki.nixos.org/wiki/Accelerated_Video_Playback) or the [Intel Graphics](https://wiki.nixos.org/wiki/Intel_Graphics) NixOS Wiki Pages.

**Comment by Atemu** [comment]
libva != libvpl

VAAPI is enabled in all ffmpeg builds.

Please also note that, while the current NixOS wiki is now "official", the articles you linked were merely copied over from the unofficial wiki. They're not necessarily correct or up to date and do not constitute official documentation.

**Comment by septem9er** [comment]
Sorry, I misspelled.

I know that these articels are only copied from the unoffical wiki. However, I still think that this should be documented _somewhere_.

However, if you don't think so that I guess this issue can be considred closed.

**Comment by Atemu** [comment]
> However, I still think that this should be documented _somewhere_.

Absolutely.

The problem is that we don't really have a place to properly document override flags yet. See https://github.com/NixOS/rfcs/pull/169 for progress on that.

Though in the mean time, we could at least document how to get specific things working and point users towards reading the source files to know which flags exist.

I don't have OneVPL working on my system, so I can't be of help here but if anyone has, it'd be great if you could expand the wiki on that topic :)

> However, if you don't think so that I guess this issue can be considred closed.

No I do think so and I also think that libvpl should be enabled in ffmpeg. I'd prefer to keep this issue open in case anyone has a solution.

**Comment by natto1784** [comment]
@Atemu Hey, I still think enabling libvpl for 6.0 and above should be the default! People with older GPUs can override it anyway for 6.0 and above.

**Comment by Atemu** [comment]
@natto1784 as explained earlier, that would imply disabling support for older iGPUs which are still quite common. We cannot do that. Overriding ffmpeg system-wide is not feasible for most users.

With https://github.com/NixOS/nixpkgs/pull/315425 you should at least be able to use intel-media-sdk via VPL though. Could you try that? I'm also not really sure what exactly it is that you want to achieve using oneVPL that you couldn't achieve using libva.

Until it is possible to enable VPL and intel-media-sdk at the same time in ffmpeg, I'd consider this issue to be resolved. This issue is not actionable on our side.

**Comment by natto1784** [comment]
Yes, well I have already got everything working with the said flags, that is a non-issue. I was just looking forward to making it the default. Well too bad that it is not feasible right now.

**Comment by Atemu** [comment]
By "said flags", did you mean enabling oneVPL via override or using oneVPL via intel-media-sdk?

I'm also still curious what the use-case is over libva.

**Comment by natto1784** [comment]
i meant simply enabling vpl, i just want to be able to get this, without rebuilding jellyfin every time, lol
```sh
nix shell --impure --expr 'with (import <nixpkgs> {}); jellyfin.override { ffmpeg = jellyfin-ffmpeg.override { ffmpeg_6-full = ffmpeg_6-full.override { withMfx = false; withVpl = true; };};}'
```

**Comment by rekh127** [comment]
@Atemu Intels blog post about the new option in FFMPEG states "Users can switch to VPL in FFMPEG, even for older Intel hardware."  They explain that they have a layer which selects either actual VPL or the older Media SDK. 
https://www.intel.com/content/www/us/en/developer/articles/technical/onevpl-in-ffmpeg-for-great-streaming-on-intel-gpus.html

Have you found this not to actually work on older intel iGPU's or was it an assumption that it wouldn't?

**Comment by Atemu** [comment]
If it's possible to use VPL in ffmpeg without removing support for widely used GPUs, I'm all for it.

So far, I have simply not seen any report of it actually working however.

**Comment by GeorgeZaher** [comment]
Hello,
I am exteremly intrested in getting ffmpeg with libvpl, please
I do not care if rest of GPUs do not work, but reather to have something work first
Best Regards
George

**Comment by Atemu** [comment]
As mentioned previously in this thread, please simply override your ffmpeg: `ffmpeg.override { withVPL = true; withMFX = false; }`

**Comment by huantianad** [comment]
Issue resurrection time!

From what I can untangle from the mess of the intel GPU stack, switching to `libvpl` in ffmpeg should *not* drop support for older hardware supported by `libmfx`, as `libvpl` should dispatch to the Media SDK when needed. This is mentioned by @PJungkamp in [this comment](https://github.com/NixOS/nixpkgs/pull/315415#issuecomment-2156459819), and I'm wondering, if someone can test this and confirm. If so, I'd say we're good to enable `libvpl` instead of `libmfx`.

**Comment by PJungkamp** [comment]
As I said, I don't have older hardware to test.

And if we were to switch, every user of ffmpeg with Intel QSV needs to add intel-media-sdk to the graphics extra packages on those older CPUs.

**Comment by huantianad** [comment]
I see, I was under the impression that we already needed to add `intel-media-sdk` to graphics extra packages.

**Comment by nyanmisaka** [comment]
@Atemu
FYI the [LibVPL](https://github.com/intel/libvpl) is just a loader that selects the appropriate runtime version automatically based on the generation of the Intel GPU - [MediaSDK runtime](https://github.com/Intel-Media-SDK/MediaSDK) for 10th Gen and older, and [VPL runtime](https://github.com/intel/vpl-gpu-rt) for 11th Gen and newer. To be compatible with both old and new GPUs, just make sure both runtimes are installed at the same time, which will not break anything.

Without VPL enabled, the `av1_qsv` hardware encoder in FFmpeg is not available.

**Comment by Atemu** [comment]
Thanks for the input @nyanmisaka.

I am now convinced that this is a good idea. I'd like to invite interested users to create a PR against staging that flips the defaults to enable VPL. I'd then need somebody with a legacy GPU to confirm that the result still works for them and then we can merge.

**Comment by justinas** [comment]
As part of https://github.com/NixOS/nixpkgs/issues/356535, I tested VPL on i3-10100 (Gen11 graphics). It works with caveats, see https://github.com/NixOS/nixpkgs/issues/356535#issuecomment-2509811952 for more info.

**Comment by otisdog8** [comment]
For what it's worth, I had to apply the workaround from https://github.com/NixOS/nixpkgs/issues/303074#issuecomment-2048868966 because otherwise jellyfin would silently fallback to software encoding for av1 (on an Arc A380). Once I enabled vpl it worked flawlessly.

**Comment by Pheoxy** [comment]
Any one got any updates on this?

My Intel Arc Battlemage B580 needs this and while I'm all for backwards compatibility I'd expect this to not be a issue just ffmpeg chooses the right backend.
